### PR TITLE
Authentication microservice

### PR DESCRIPTION
**Pull request by PaulReeve717**
PR for our custom Authentication microservice

### CODE REVIEW

File: ochap-authentication/src/main/java/com/example/ochapauthentication/entities/User.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package com.example.ochapauthentication.entities;
+
+import lombok.*;
+
+import javax.persistence.*;
+
+@Data
+@Entity
+@Table(name = "user_table")
```
Comment (roosemberth) [inline ochap-authentication/src/main/java/com/example/ochapauthentication/entities/User.java:9]:
```suggestion
@Table(name = "users")
```

File: ochap-authentication/src/main/java/com/example/ochapauthentication/entities/User.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Integer id;
+
+    @Column(nullable = false, unique = true)
+    private String username;
+
+    @Column(nullable = false)
+    private byte[] password;
```
Comment (roosemberth) [inline ochap-authentication/src/main/java/com/example/ochapauthentication/entities/User.java:19]: This is not storing the password, but rather the hash of the password.

```suggestion
    private byte[] passwordHash;
```

File: ochap-authentication/src/main/java/com/example/ochapauthentication/jwt/JwtTokenUtil.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    @Value("${jwt.secret}")
+    private String secret;
+
+    public String generateToken(AccountInfoDTO userDetails) {
+        Map<String, Object> claims = new HashMap<>();
+        claims.put("role", userDetails.getRole());
+        return doGenerateToken(claims, userDetails.getUsername());
+    }
+
+    private String doGenerateToken(Map<String, Object> claims, String subject) {
```
Comment (roosemberth) [inline ochap-authentication/src/main/java/com/example/ochapauthentication/jwt/JwtTokenUtil.java:29]:
```suggestion
    private String generateJwtWithClaims(String subject, Map<String, Object> claims) {
```

Nitpick: Claims are associated to the subject, thus it is more intuitive to specify the subject first.

File: ochap-authentication/src/main/java/com/example/ochapauthentication/routes/AuthenticationController.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        this.roleRepository = roleRepository;
+        this.objectMapper = new ObjectMapper();
+    }
+
+    @Autowired
+    private JwtTokenUtil jwt;
+
+    @PostMapping("/auth/login")
+    @ResponseBody
+    ResponseEntity<String> loginAutentication(@RequestBody AuthLoginCommand credentials) throws JsonProcessingException {
```
Comment (roosemberth) [inline ochap-authentication/src/main/java/com/example/ochapauthentication/routes/AuthenticationController.java:54]:
```suggestion
    ResponseEntity<Object> loginAutentication(@RequestBody AuthLoginCommand credentials) throws JsonProcessingException {
```

Why not let spring to the marshalling for you?

File: ochap-authentication/src/main/java/com/example/ochapauthentication/routes/AuthenticationController.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        AccountInfoDTO accountInfo = new AccountInfoDTO(u.getId(), u.getUsername(), u.getRole().getName());
+
+        TokenDTO tokenDTO = new TokenDTO(jwt.generateToken(accountInfo), accountInfo);
+
+        return new ResponseEntity<>(objectMapper.writeValueAsString(tokenDTO), HttpStatus.OK);
+    }
+
+    @PostMapping(value = "/accounts/register",  produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseBody
+    ResponseEntity<String> registerAutentication(@RequestBody AccountRegisterCommand credentials) throws JsonProcessingException {
```
Comment (roosemberth) [inline ochap-authentication/src/main/java/com/example/ochapauthentication/routes/AuthenticationController.java:76]:
```suggestion
    ResponseEntity<String> registerAutentication(@RequestBody AccountRegisterCommand credentials) throws JsonProcessingException {
```

Why not let spring to the marshalling for you?

File: ochap-authentication/src/main/java/com/example/ochapauthentication/routes/AuthenticationController.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        if(isError){
+            ErrorsDTO errors = new ErrorsDTO(errorList);
+            return new ResponseEntity<>(objectMapper.writeValueAsString(errors), HttpStatus.UNPROCESSABLE_ENTITY);
+        }
+
+        if (userRepository.getUserByUsername(credentials.getUsername()).isPresent()) {
+            com.example.ochapauthentication.dto.Error alreadyExisting =
+            new com.example.ochapauthentication.dto.Error("The username already exist");
```
Comment (roosemberth) [inline ochap-authentication/src/main/java/com/example/ochapauthentication/routes/AuthenticationController.java:97]:
```suggestion
            new com.example.ochapauthentication.dto.Error("A user with the specified username already exists.");
```

The username identifies a user, but is not an entity on its own.

File: ochap-authentication/pom.xml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        </dependency>
+
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.springframework.boot</groupId>
+                <artifactId>spring-boot-maven-plugin</artifactId>
+            </plugin>
```
Comment (roosemberth) [inline ochap-authentication/pom.xml:89]: Missing formatter plugin:

```suggestion
            </plugin>
			<plugin>
				<groupId>net.revelc.code.formatter</groupId>
				<artifactId>formatter-maven-plugin</artifactId>
				<version>2.16.0</version>
				<configuration>
					<configFile>${project.basedir}/../eclipse-java-google-style.xml</configFile>
				</configuration>
			</plugin>
```

File: ochap-authentication/pom.xml [hunk 8]
```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (roosemberth) [inline ochap-authentication/pom.xml:1]: This module should be added to the parent pom aswell.

Comment (roosemberth) [inline ochap-authentication/pom.xml:1]: This module is missing a README.

**Review by roosemberth — CHANGES_REQUESTED** [review]
I've left some small comments, but overall looks good, great work!

File: ochap-authentication/src/main/java/com/example/ochapauthentication/dto/Error.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package com.example.ochapauthentication.dto;
+
+import lombok.Data;
+
+@Data
+public class Error {
```
[the remark this answers is not in the document]
Comment (Klab93) [reply ochap-authentication/src/main/java/com/example/ochapauthentication/dto/Error.java:6]: AuhtenticationController line 60

File: ochap-authentication/README.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+Also after the database schema creation after the first time the server is booted, there's the need 
+to manually insert the two default roles in the `role` table running the following sql lines:
+```sql
+INSERT INTO public.role(
+	name)
+	VALUES ('user');
+INSERT INTO public.role(
+	name)
+	VALUES ('admin');
+```
```
Comment (roosemberth) [inline ochap-authentication/README.md:25]: This section is outdated.

**Review by roosemberth — CHANGES_REQUESTED** [review]
Almost there, just a little update on the README!

**Review by roosemberth — APPROVED** [review]
Amazing! Thanks for your contribution!

**Pull request merged**